---
title: "Drosophila Metabarcoding"
subtitle: "Statistical analysis"
author: "Alexander Piper"
date: "`r Sys.Date()`"
output:
  html_document:
    highlighter: null
    theme: "flatly"
    code_download: true
    toc: true
    toc_float: 
      collapsed: false
      smooth_scroll: true
    df_print: paged    
  pdf_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Knitr global setup - change eval to true to run code
library(knitr)
knitr::opts_chunk$set(echo = TRUE, eval=FALSE, message=FALSE,error=FALSE, fig.show = "hold", fig.keep = "all")
opts_chunk$set(dev = 'png')
```

## Load packages

```{r load packages}
# Load the Renv containing package versions
#renv::restore()

#Set required packages
.cran_packages <- c("tidyverse",
                    "tidymodels",
                    "probably",
                    "patchwork", 
                    "vegan", 
                    "seqinr",
                    "ape", 
                    "RColorBrewer",
                    "castor", 
                    "picante",
                    "phytools",
                    "ggrepel",
                    "devtools",
                    "ggnewscale",
                    "ggpubr",
                    "ggdist",
                    "mvabund")
.bioc_packages <- c("dada2",
                    "phyloseq", 
                    "ggtree",
                    "DECIPHER",
                    "Biostrings",
                    "ShortRead", 
                    "philr",
                    "ALDEx2")

.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(.bioc_packages[!.inst], ask = F)
}

sapply(c(.cran_packages,.bioc_packages), require, character.only = TRUE)

# Github packages
devtools::install_github("alexpiper/taxreturn")
devtools::install_github("alexpiper/seqateurs")
devtools::install_github("mikemc/speedyseq")
devtools::install_github('ggloor/CoDaSeq/CoDaSeq')
devtools::install_github("mikemc/metacal")
devtools::install_github("adw96/breakaway")

library(speedyseq)
library(taxreturn)
library(seqateurs)
library(CoDaSeq)
library(metacal)
library(breakaway)

#Source themes
source('R/themes.R')
```

## Read in phyloseq object

```{r phyloseq}
ps1 <- readRDS("output/rds/ps1.rds")
```

## Replicate observations and overlap

```{r Replicates}
# Abundance counts for different replicates
# Load in all samples that were expected
exp_samples <- read_csv("data/HLVKYDMXX/SampleSheet.csv", skip = 19) %>%
  dplyr::select(extract_id = Sample_Name, sample_name = Sample_Name) %>%
  group_by(extract_id) %>%
  #group_split() %>%
  group_modify(~{
    .x %>%
      dplyr::slice(rep(1:n(), each=3)) %>%
      mutate(amp_rep = row_number()) %>%
      mutate(sample_id = paste0(sample_name, "_", amp_rep))
  }) %>%
  ungroup() %>%
  mutate(sample_name = extract_id %>%
           str_remove("-ex[0-9]$"))%>%
  mutate(type = case_when(
    str_detect(sample_id, "D[0-9][0-9][0-9]M|D[0-9][0-9][0-9][0-9]M|DM[0-9]")  ~ "DrosMock",
    str_detect(sample_id, "SPD")  ~ "SPD",
    str_detect(sample_id, "ACV")  ~ "ACV",
    str_detect(sample_id, "DC")  ~ "DC",
    str_detect(sample_id, "Sach")  ~ "DC",
    str_detect(sample_id, "FF")  ~ "FF",
    str_detect(sample_id, "NTC")  ~ "NTC",
    str_detect(sample_id, "DLarv")  ~ "DrosLarv",
    str_detect(sample_id, "POS|SynMock")  ~ "POS",
    str_detect(sample_id, "extblank|BLANK")  ~ "Extblank",
    str_detect(sample_id, "pcrblank")  ~ "PCRblank",
    str_detect(sample_id, "CT")  ~ "CarpTrap",
    str_detect(sample_id, "CM[0-9]")  ~ "CarpMock",
    str_detect(sample_id, "CML[0-9]")  ~ "CarpLarval"
  )) %>%
  filter(type %in% c("DrosMock", "DrosLarv", "ACV", "DC", "SPD", "FF", "POS", "SPD"))
  
  
# Heatmap of abundance of pcr replicates
rep_data <- exp_samples %>%
  left_join(ps1 %>% 
  speedyseq::psmelt() %>%
    filter(fcid=="HLVKYDMXX", Abundance > 0) %>%
    group_by(sample_name, extract_id, sample_id, amp_rep, fcid) %>%
    summarise(Abundance = sum(Abundance), n_spp = n_distinct(Species)) %>%
    ungroup()) %>%
  filter(!str_detect(sample_name, "_DM1$|_DM2$|_DM3$|_DM4$|_DM5$")) %>%
  mutate(extrep = extract_id %>% str_extract("ex.*$")) %>%
  mutate(type = type %>% 
           str_replace("DrosMock", "Adult Mock") %>%
           str_replace("DrosLarv", "Larval Mock") %>%
           factor(levels=c("POS", "Adult Mock", "Larval Mock","FF","SPD", "DC", "ACV"))) %>%
  mutate(sample_name = sample_name %>%
           str_remove("HLVKYDMXX_") %>%
           str_replace("^D2","D02") %>%
           str_replace("^D5","D05")) %>%
  #filter(!(extrep=="ex1" & amp_rep == 2)) %>%
  mutate(amp_rep = factor(amp_rep)) %>%
  mutate(extrep = extrep %>% str_replace("ex", "Extraction Rep "))

# Summarise of replicate success
rep_data %>%
  ungroup() %>%
  mutate(success = case_when(
    Abundance > 0 ~ TRUE,
    Abundance == 0 | is.na(Abundance) ~ FALSE
  )) %>%
  group_by(type) %>%
  summarise(sum=sum(success), total = n()) %>%
  mutate(proportion = sum / total)

# Summary of sample success
rep_data %>%
  ungroup() %>%
  group_by(sample_name, type) %>%
  summarise(Abundance = sum(Abundance, na.rm = TRUE))%>%
  ungroup() %>%
  mutate(success = case_when(
    Abundance > 0 ~ TRUE,
    Abundance == 0 | is.na(Abundance) ~ FALSE
  )) %>%
  group_by(type) %>%
  summarise(sum=sum(success), total = n()) %>%
  mutate(proportion = sum / total)


gg.repmap <- rep_data %>%
   ggplot(aes(x = amp_rep, y = sample_name, fill = n_spp)) +
  geom_tile()+
  scale_fill_viridis_c(breaks=c(1,5,10,15,20,24), labels= c(1,5,10,15,20,24))+
  facet_grid(extrep~type, scales="free", space="free", drop=TRUE)+
  base_theme+
  scale_x_discrete(expand=c(0,0))+
  scale_y_discrete(expand=c(0,0))+
  coord_flip()+
  theme(axis.text.x = element_blank(),
        legend.position = "top")+
  labs(x = "PCR Replicate",
       y = NULL,
       fill = "N Species")

gg.repmap

# What species are different between samples
diff_between_samples <- ps1 %>% 
  speedyseq::psmelt() %>%
  filter(fcid=="HLVKYDMXX", Abundance > 0) %>%
  dplyr::select(OTU, sample_name, sample_id, extract_id, Species, Abundance) %>%
  distinct() %>%
  group_by(sample_id) %>%
  mutate_at(vars(Abundance), ~ . / sum(., na.rm = TRUE) ) %>%
  ungroup()%>%
  group_by(sample_name) %>%
  mutate(n = n_distinct(sample_id)) %>%
  ungroup()%>%
  group_by(sample_name, Species) %>%
  mutate(diff = n_distinct(sample_id)) %>%
  ungroup() %>%
  mutate(perc = diff / n)

diff_between_samples %>% 
  ggplot(aes(x = Abundance, y=perc)) +
  geom_point(position=position_jitter(width=0.01, height=0.01), alpha=0.1) + 
  scale_x_log10(labels=scales::percent)+
  geom_smooth(method="lm") +
  base_theme

broom::tidy(lm(formula = perc ~ Abundance, data=diff_between_samples))

# Get mean pairwise distance within all replicates in a sample, or number of taxa detected?
rep_distance <- ps1 %>%
  subset_samples(fcid=="HLVKYDMXX") %>%
  speedyseq::tax_glom(taxrank = "Species") %>%
  otu_table() %>%
  as("matrix") %>%
  as_tibble(rownames = "sample_id") %>%
  #filter(!str_detect(sample_id, "BLANK|SynMock|POS")) %>%
  left_join(sample_data(ps1) %>%
              as("data.frame") %>%
              as_tibble()%>%
              dplyr::filter(fcid=="HLVKYDMXX") %>%
              dplyr::select(sample_name, sample_id)  %>%
              distinct()) %>%
  group_by(sample_name) %>%
  add_tally() %>%
  filter(n > 1) %>%
  dplyr::select(-n) %>%
  group_modify(~ {
      .x %>% 
        column_to_rownames("sample_id")%>%
        dplyr::select(-any_of(c("sample_id", "sample_name", "extract_id", "type", "fcid"))) %>%
        dplyr::select(where(~ !(sum(.)==0))) %>%
        as.matrix() %>%
        vegdist(method="jac", binary = TRUE) %>%
        as.matrix() %>%
        as_tibble(rownames="sample_1") %>%
        pivot_longer(-sample_1,
                     names_to = "sample_2",
                     values_to = "dist")%>%
        dplyr::filter(dist > 0)
    })

gg.repjac <- rep_distance %>%
  mutate(sample_name = sample_name %>%
           str_remove("HLVKYDMXX_") %>%
           str_replace("^D2","D02") %>%
           str_replace("^D5","D05")) %>%
  right_join(rep_data %>% dplyr::select(sample_name, type) %>% distinct()) %>%
  mutate(type = type %>%
           factor(levels=c("POS", "Adult Mock", "Larval Mock","FF","SPD", "DC", "ACV"))) %>%
  #filter(!is.na(type)) %>%
  ggplot(aes(x = sample_name, y = 1-dist, colour=1-dist))+
  geom_boxplot(outlier.color = NA)  +
  geom_point(alpha=0.5, position= position_jitter(width=0.2))+
  scale_colour_viridis_c( )+
  coord_cartesian(ylim = c(0,1))+
  scale_y_continuous(labels = scales::percent)+
  base_theme+
  theme(strip.text.x = element_blank())+
  labs(x = "Sample",
       y = "Pairwise Jaccard similarity")+
  facet_grid(.~type, scales="free", space="free", drop=TRUE)

# pairwise distance as function of pairwise difference in sequencing depth
gg.jacpw <- rep_distance %>%
  mutate(sample_name = sample_name %>%
           str_remove("HLVKYDMXX_") %>%
           str_replace("^D2","D02") %>%
           str_replace("^D5","D05")) %>%
  left_join(rep_data %>%
               dplyr::select(sample_1 = sample_id, abundance_1 = Abundance) %>%
               distinct()) %>%
  left_join(rep_data %>%
               dplyr::select(sample_2 = sample_id, abundance_2 = Abundance) %>%
               distinct()) %>%
  mutate(diff_abund = abs(abundance_2 -abundance_1)) %>%
  ggplot(aes(x = diff_abund, y = 1-dist, colour=1-dist)) +
  geom_point(alpha=0.2) +
  geom_smooth(method="lm") +
  stat_cor(label.y = 0.95, label.x = 10,size=3,
           aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))) +
  scale_y_continuous(labels = scales::percent, position="right")+
  scale_x_continuous(labels = scales::label_number_si())+
  scale_colour_viridis_c( )+
  base_theme +
  theme(axis.text.x = element_text(angle=0))+
  labs(x = "Pairwise difference in seq depth",
       y = "Pairwise Jaccard Similarity")+
  coord_cartesian(ylim = c(0,1))
  

gg.jacpw

# Anova of difference
rep_dist_types <- rep_distance %>%
  left_join(sample_data(ps1) %>%
              as("matrix") %>%
              as.data.frame() %>%
              dplyr::select(sample_name, type) %>%
              distinct()) %>%
  dplyr::filter(!(sample_1 == sample_2)) %>%
  group_by(type) %>%
  add_tally() %>%
  filter(!type %in% c("ACV", "POS"))

rep_dist_types %>%
  ggplot(aes(x = type, y=dist)) +
  geom_boxplot() +
  geom_point(alpha=0.1)

report::report(aov(dist ~type, data=rep_dist_types))
broom::tidy(TukeyHSD(aov(dist ~type, data=rep_dist_types))) %>%
  mutate(signif = case_when(
    adj.p.value < 0.05 ~ TRUE,
    TRUE ~ FALSE
  )) %>%
  tidyr::separate(contrast, into=c("contrast_1", "contrast_2"), sep="-")


# Mean pairwise distance between pcr replicates within extraction
mpd_pcr <- ps1 %>%
  subset_samples(fcid=="HLVKYDMXX") %>%
  speedyseq::tax_glom(taxrank = "Species") %>%
  otu_table() %>%
  as("matrix") %>%
  as_tibble(rownames = "sample_id") %>%
  filter(!str_detect(sample_id, "BLANK|SynMock|POS")) %>%
  left_join(sample_data(ps1) %>%
              as("data.frame") %>%
              as_tibble()%>%
              dplyr::filter(fcid=="HLVKYDMXX") %>%
              dplyr::select(sample_id, sample_name, extract_id)  %>%
              distinct()) %>%
  group_by(sample_name, extract_id) %>%
  add_tally()  %>%
  filter(n > 1) %>%
  dplyr::select(-n) %>%
  group_modify(~ {
      .x %>% 
        dplyr::select(-any_of(c("sample_id", "sample_name", "extract_id", "fcid"))) %>%
        dplyr::select(where(~ !(sum(.)==0))) %>%
        as.matrix() %>%
        vegdist(method="jac", binary = TRUE) %>%
        as.matrix() %>%
        as_tibble()%>%
        pivot_longer(everything(),
                     names_to = "sample",
                     values_to = "dist")%>%
        dplyr::filter(dist > 0) %>%
        summarise(dist = mean(dist), n = n_distinct(sample)) 
    })

# Mean pairwise distance of extraction reps within samples
mpd_extraction <- ps1 %>%
  subset_samples(fcid=="HLVKYDMXX") %>%
  merge_samples(group = "extract_id", fun="sum") %>%
  speedyseq::tax_glom(taxrank = "Species") %>%
  otu_table() %>%
  as("matrix") %>%
  as_tibble(rownames = "extract_id") %>%
  filter(!str_detect(extract_id, "BLANK|SynMock|POS")) %>%
  left_join(sample_data(ps1) %>%
              as("data.frame") %>%
              as_tibble()%>%
              dplyr::filter(fcid=="HLVKYDMXX") %>%
              dplyr::select(sample_name, extract_id)  %>%
              distinct()) %>%
  group_by(sample_name) %>%
  add_tally() %>%
  filter(n > 1) %>%
  dplyr::select(-n) %>%
  group_modify(~ {
      .x %>% 
        dplyr::select(-any_of(c("sample_id", "sample_name", "extract_id", "type", "fcid"))) %>%
        dplyr::select(where(~ !(sum(.)==0))) %>%
        as.matrix() %>%
        vegdist(method="jac", binary = TRUE) %>%
        as.matrix() %>%
        as_tibble()%>%
        pivot_longer(everything(),
                     names_to = "sample",
                     values_to = "dist")%>%
        dplyr::filter(dist > 0) %>%
        summarise(dist = mean(dist), n = n_distinct(sample)) 
    })


# Mean pairwise distance within collection methods
mpd_traps <- ps1 %>%
  subset_samples(fcid=="HLVKYDMXX") %>%
  merge_samples(group = "sample_name", fun="sum") %>%
  speedyseq::tax_glom(taxrank = "Species") %>%
  otu_table() %>%
  as("matrix") %>%  
  as_tibble(rownames = "sample_name") %>%
  filter(!str_detect(sample_name, "BLANK|SynMock|POS")) %>%
  left_join(sample_data(ps1) %>%
              as("data.frame") %>%
              as_tibble()%>%
              dplyr::filter(fcid=="HLVKYDMXX") %>%
              dplyr::select(sample_name, type)  %>%
              distinct()) %>%
  group_by(type) %>%
  add_tally() %>%
  filter(n > 1) %>%
  dplyr::select(-n) %>%
  group_modify(~ {
      .x %>% 
        dplyr::select(-any_of(c("sample_id", "sample_name", "extract_id", "type", "fcid"))) %>%
        dplyr::select(where(~ !(sum(.)==0))) %>%
        as.matrix() %>%
        vegdist(method="jac", binary = TRUE) %>%
        as.matrix() %>%
        as_tibble()%>%
        pivot_longer(everything(),
                     names_to = "sample",
                     values_to = "dist")%>%
        dplyr::filter(dist > 0)  %>%
        summarise(dist = mean(dist), n = n_distinct(sample)) 
    }) %>%
  mutate(enviro = type)

# Plot together:
gg.mpd <- do.call("list", mget(grep("mpd_",names(.GlobalEnv),value=TRUE))) %>%
  bind_rows(.id="type") %>%
  ungroup %>%
  dplyr::select(sample_name, type, dist, enviro) %>%
  mutate(type = type %>% 
           str_remove("^.+_") %>%
           str_replace("orchard", "Location") %>%
           str_replace("traps", "Biological reps") %>%
           str_replace("extraction", "Extraction reps") %>%
           str_replace("pcr", "PCR reps") %>%
           factor(levels = c("PCR reps", "Extraction reps", "Biological reps", "Location"))
         ) %>%
  mutate(Community = case_when(
    str_detect(sample_name, "_D[0-9]|_DM|_DL") ~ "Mock",
    enviro %in% c("Mock", "DrosLarv", "DrosMock") ~ "Mock",
    TRUE ~ "Trap"
  )) %>%
  ggplot(aes(x = type, y = 1-dist, fill=Community)) + 
  geom_point(position=position_jitter(width = 0.1),size=2, shape=21, alpha=0.8, colour="black")+
  geom_boxplot(alpha=0.3, colour="black", fill=NA, outlier.colour = NA, show.legend = FALSE) +
  base_theme + 
  theme(legend.position = "top")+
  coord_cartesian(ylim=c(0,1))+
  scale_y_continuous(labels = scales::percent, position = "right")+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust=0.5))+
  #coord_flip(ylim=c(0,1))+
  scale_fill_viridis_d()+
  scale_colour_viridis_d()+
  #scale_colour_brewer(palette= "Set1") +
  labs(x = NULL,
       y = "Mean Pairwise Jaccard similarity")

gg.mpd 

# Replicate figure
gg.repfiga <- gg.repmap / gg.repjac + plot_layout(heights = c(2,1))
gg.repfigb <- gg.mpd / gg.jacpw  + plot_layout(heights = c(2,1))
gg.rep_multifig <-  gg.repfiga - gg.repfigb + plot_layout(widths = c(5,1)) + plot_annotation(tag_levels = "A") 
gg.rep_multifig

pdf(file="fig/Fig2_replicates.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg.rep_multifig)
try(dev.off(), silent=TRUE)
```


## Train detection model

We will filter index switching using a logistic regression model trained on the mocks

```{r index switch removal}
# Subset to mocks
ps_model <- ps1

tax_table(ps_model)[,8] <- tax_table(ps_model)[,8]%>%
  str_replace("Drosophila albomicans", "Drosophila immigrans") %>%
  str_replace("Drosophila nasuta", "Drosophila immigrans") %>%
  str_replace("Drosophila hypocausta", "Drosophila immigrans") %>%
  str_replace("Drosophila pulaua", "Drosophila immigrans") %>%
  str_replace("Drosophila kohkoa", "Drosophila immigrans") %>%
  str_replace("Drosophila rubida", "Drosophila immigrans") %>%
  str_replace("Drosophila sulfurigaster", "Drosophila immigrans") %>%
  str_replace(" ", "_")
  

ps_model <- ps_model %>%
  speedyseq::tax_glom(taxrank = "Species") %>%
  subset_samples(fcid %in% c("CB3DR", "HLVKYDMXX")) %>%
  filter_taxa( function(x) mean(x) > 0, TRUE) 

sample_data(ps_model)$sample_name <- sample_data(ps_model)$sample_name  %>%
  str_replace("DM1$", "D100M1") %>%
  str_replace("DM2$", "D100M2") %>%
  str_replace("DM3$", "D100M3") %>%
  str_replace("DM4$", "D100M4") %>%
  str_replace("DM5$", "D100M5")

sample_data(ps_model)$sample_id <- sample_data(ps_model)$sample_id  %>%
  str_replace("DM1$", "D100M1") %>%
  str_replace("DM2$", "D100M2") %>%
  str_replace("DM3$", "D100M3") %>%
  str_replace("DM4$", "D100M4") %>%
  str_replace("DM5$", "D100M5")

exp <- read_csv("sample_data/expected_quant_merged.csv") %>%
  dplyr::rename(sample_name = Sample) %>%
  pivot_longer(-sample_name,
               names_to= "taxon",
               values_to= "expected") %>%
  mutate(taxon = str_replace(taxon, pattern=" ",replacement="_"),
         sample_name = str_remove(sample_name, "-exp*.$")) %>%
  dplyr::filter(!is.na(sample_name),
  !str_detect(sample_name, "CM[0-9]|CT[0-9]|CML[0-9]|Syn"),
  expected > 0) %>%
  distinct() %>%
  mutate(expected_RA = expected) %>%
  group_by(sample_name) %>%
  mutate_at(vars(expected_RA), ~ . / sum(., na.rm = TRUE) ) %>%
  ungroup()


# Merge replicates and get observed
sam1 <- speedyseq::psmelt(ps_model) %>%
  janitor::clean_names() %>%
  mutate(taxon = species) %>%
  filter(fcid %in% c("CB3DR", "HLVKYDMXX"))  %>%
  group_by(sample_id,fcid, pcr_primers)  %>%
  dplyr::mutate(abundance_RA = abundance, total_seq = sum(abundance)) %>%
  #mutate(abundance_RA = abundance_RA + 0.5) %>% # Add pseudocount
  mutate_at(vars(abundance_RA), ~ . / sum(.) ) %>%
  group_by(sample_id, fcid, pcr_primers)  %>%
  mutate(abundance_clr = abundance_RA) %>%
  mutate_at(vars(abundance_clr ), ~metacal::clr(.) ) %>%
  ungroup() %>%
  dplyr::select(otu, sample_name, extract_id, sample_id, taxon, pcr_primers,fcid, material_type = type, abundance, abundance_RA, abundance_clr,total_seq) %>%
  distinct() 

repcounts <- sam1 %>%
 # filter(abundance > 0) %>% 
  group_by(sample_name, otu, fcid, pcr_primers) %>%
  mutate(
    n_extract = case_when( # Need this to only count reps they were detected in!
    abundance > 0 ~ extract_id,
    TRUE ~ as.character(NA)
  ),
    n_rep = case_when(
    abundance > 0 ~ sample_id,
    TRUE ~ as.character(NA)
  )) %>%
  summarise(n_extract = n_distinct(n_extract, na.rm = TRUE), n_rep = n_distinct(n_rep, na.rm = TRUE)) 

sam <- sam1%>%
  left_join(repcounts)%>%
  mutate(taxon = taxon %>% str_replace(" ", "_")) %>%
  mutate(sample_id = sample_name,
         sample_name = sample_name %>%      
         str_remove_all("BF1-BR1-|SauronS878-HexCOIR4-|fwhF2-HexCOIR4-|fwhF2-fwhR2n-") %>%
         str_remove_all("HLVKYDMXX_|CK3HD_|CB3DR_|CJKFJ_")) %>%
  group_by(otu, sample_name, sample_id, taxon, pcr_primers, fcid, material_type, n_rep, n_extract) %>%
  summarise(abundance = sum(abundance), abundance_RA = mean(abundance_RA), abundance_clr = mean(abundance_clr), total_seq = sum(total_seq)) %>%   
  ungroup() 

#Create a mocks table to train the model on
model_df <- sam %>%
  mutate(type = case_when( 
         (sample_name %in% exp$sample_name) &
         (material_type == "DrosMock") & !str_detect(taxon, "__") ~ "train",
         TRUE ~ "test")) %>%
  left_join(exp, by = c("sample_name","taxon"))  %>%
  left_join(exp %>%
              group_by(sample_name) %>%
              mutate(total_abund = sum(expected))%>%
              ungroup() %>%
              dplyr::select(sample_name, total_abund) %>%
              distinct(), by="sample_name") %>%
  mutate(outcome = case_when(
    abundance > 0 & is.na(expected) & type== "train" ~ 0, #False positive - only count for mocks
    abundance > 0 & !is.na(expected) ~ 1, # True positive
    abundance == 0 & expected > 0 ~ 1, # False negative
    abundance ==0 & is.na(expected) & type== "train" ~ 0 # True negative - only count for mocks
  )) %>%
  dplyr::select(fcid, sample_id, sample_name, otu, taxon, abundance, abundance_clr, abundance_RA,pcr_primers, outcome, n_extract, n_rep, total_seq, total_abund, expected, expected_RA) %>%
 # mutate(outcome_all = outcome) %>%
  mutate(outcome = case_when(
    outcome == 0 & abundance_RA > 0.01 ~ as.numeric(NA), #Deal with obvious misannotated expecteds
    outcome == 1 & abundance == 0 ~ as.numeric(NA), #Deal with total dropouts
    !pcr_primers == "fwhF2-fwhR2n" ~ as.numeric(NA), #Dont use the other primers
    TRUE ~ outcome
  )) %>%
  #filter(abundance > 0) %>%
  #filter(!is.na(outcome)) %>%
  mutate(outcome = factor(outcome, levels=c("0", "1")))

# Get just known outcomes
mock_dat <- model_df %>%
  filter(!is.na(outcome)) 
  #filter(!is.na(outcome), abundance > 0) 

# Create dataset split
set.seed(seed = 42) 

mock_split <- initial_split(mock_dat,strata=outcome, prop=0.8)
mock_train <- training(mock_split)
mock_test <- testing(mock_split)

# Cross validations
mock_cv <- vfold_cv(mock_train, v = 10)

# Create modelling recipe
switch_recipe <- recipe(outcome ~ abundance_RA + fcid + n_rep + n_extract, data = mock_train) %>%
  step_novel(all_nominal(), -all_outcomes()) %>% 
  step_dummy(fcid, one_hot = FALSE) %>% 
  themis::step_upsample(outcome) %>%
  step_zv(all_predictors()) %>%
  step_log(abundance_RA, base=10, offset = 0.00001) %>%
  step_normalize(abundance_RA, n_rep, n_extract)

juice(prep(switch_recipe))

# Define model
glmnet_spec <- logistic_reg() %>%  #penalty = tune(), mixture = tune()
  set_mode("classification") %>% 
  set_engine("glm") 

# Create workflow
glmnet_wf <- workflow() %>%
  add_recipe(switch_recipe) %>%
  add_model(glmnet_spec)

# See how it went on the training dataset
predictions_glm <- mock_train %>%
  nest(everything()) %>%
  dplyr::rename(train = data) %>%
  bind_cols(mock_test %>%
              nest(everything()) %>%
              dplyr::rename(test = data)) %>%
  mutate(model_obj = purrr::map(train, ~fit(glmnet_wf, .x)),
         pred_train = purrr::map2(model_obj, train, ~predict(.x, .y, type = "prob")),
         pred_test = purrr::map2(model_obj, test, ~predict(.x, .y, type = "prob")),
         vip = purrr::map(model_obj ,~{
            .x %>%
            pull_workflow_fit() %>%
            vip::vi()
         }),
         coef_info = purrr::map(model_obj,~{
           .x %>% 
             pull_workflow_fit() %>%
             broom::tidy()
         })
  )

# Confusion matrix
gg.conf_matrix <- predictions_glm %>%
  unnest(test, pred_test) %>%
  dplyr::select(-where(~is.list(.x))) %>%
  mutate(.pred = factor(make_two_class_pred(.pred_0, levels(outcome), threshold = .5)))%>%
  mutate(.pred = factor(.pred, levels = levels(outcome))) %>%
  conf_mat(outcome, .pred) %>%
  pluck(1) %>%
  as_tibble() %>%
  ggplot(aes(Prediction, Truth, alpha = n)) +
  geom_tile(show.legend = FALSE) +
  geom_text(aes(label = n), colour = "white", alpha = 1, size = 8)

gg.conf_matrix

# Calculate performance over number of predictors
threshold_data <- predictions_glm %>%
  unnest(train, pred_train) %>%
  dplyr::select(-where(~is.list(.x))) %>%
  threshold_perf(outcome, .pred_1, thresholds = seq(0, 1, by = 0.0025)) %>%
  filter(!.metric == "distance") %>%
  mutate(.metric = .metric %>% 
           str_replace("sens", "Sensitivity") %>%
           str_replace("spec", "Specificity") %>%
           str_replace("j_index", "J Index"))

max_j_index_threshold <- threshold_data %>%
  filter(.metric == "J Index") %>%
  filter(.estimate == max(.estimate)) %>%
  summarise(upper = max(.threshold), lower = min(.threshold))

gg_thresholds <- threshold_data %>%
  mutate(.metric = factor(.metric, levels = c("Sensitivity","Specificity","J Index" ))) %>%
  ggplot(aes(x = .threshold, y = .estimate, color = .metric)) +
    geom_line(size=1) +
    base_theme +
    scale_color_viridis_d(end = 0.9) +
    scale_alpha_manual(values = c(.4, 1), guide = "none") +
    geom_rect(data = max_j_index_threshold, aes(xmin = lower, xmax=upper,
                 ymin=0, ymax=1), alpha = .3, fill = "grey30", inherit.aes = FALSE) +
    #geom_vline(xintercept = max_j_index_threshold, alpha = .6, color = "grey30")+
    facet_grid(.metric~.) +
    labs(
      x = "Confidence threshold\n(above this value is considered a detection)",
      y = "Metric Estimate",
      title = "Balancing performance by varying the confidence threshold"
    ) 

gg_thresholds

pdf(file="fig/supplementary/logistic_prob_thresholds.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg_thresholds)
try(dev.off(), silent=TRUE)

# Get coefficients
predictions_glm  %>%
  dplyr::select(coef_info) %>%
  unnest() %>%
  mutate(signif = case_when(
    p.value < 0.001 ~ TRUE,
    p.value > 0.001 ~ FALSE
  )) %>%
 # mutate(estimate = exp(estimate)) %>%
  ggplot(aes(x = estimate, y = term, colour=signif)) +
  geom_pointrange(aes(x = estimate, xmax=estimate+std.error, xmin=estimate-std.error ), size=1) +
  labs(y = NULL) +
  base_theme + 
  geom_vline(xintercept=0) + 
  theme(legend.position = "right")
  
# Get importance of predictors
predictions_glm  %>%
  dplyr::select(vip) %>%
  unnest() %>%
  group_by(Sign) %>%
  ggplot(aes(x = Importance, y = Variable, fill = Sign)) +
  geom_col() +
  labs(y = NULL)

# Check for disagreements within mock
mock_mismatch <- predictions_glm %>%
  unnest(test, pred_test) %>%
  dplyr::select(-where(~is.list(.x))) %>%
  mutate(.pred = factor(make_two_class_pred(.pred_0, levels(outcome), threshold = .5)))%>%
  #dplyr::select(.pred, abundance_clr,abundance,  taxon, outcome)  %>%
  mutate(diff = !.pred == outcome)

# plot curve
roc_data <- predictions_glm %>%
   unnest(test, pred_test) %>%
  roc_curve(truth = outcome, .pred_0) 
gg.roc_curve <- roc_data %>%  
  ggplot(aes(x = 1 - specificity, y = sensitivity)) +
  geom_path() +
  geom_abline(lty = 3) + 
  coord_equal()

gg.roc_curve

# Calculate metrics
multimetric <- metric_set(accuracy, bal_accuracy, sens, yardstick::spec, precision, recall, ppv, npv)
predictions_glm %>%
  unnest(test, pred_test) %>%
  dplyr::select(-where(~is.list(.x))) %>%
  mutate(.pred = make_two_class_pred(.pred_0, levels(outcome), threshold = .5)) %>%
  multimetric(truth = outcome, estimate = factor(.pred, levels= levels(outcome)))  
```

# merge replicates and remove index switching

```{r merge replicates}
# Merge replicates
ps.merged <- ps_model %>%
    merge_samples(group = "sample_name", fun="sum")

##This loses the sample metadata - Need to add it agian
sample_data(ps.merged) <- sample_data(ps_model) %>%
  as("data.frame")  %>%
  mutate(sample_id = sample_name) %>%
  mutate(sample_name = sample_name %>%
           str_remove_all("BF1-BR1-|SauronS878-HexCOIR4-|fwhF2-HexCOIR4-|fwhF2-fwhR2n-") %>%
           str_remove("HLVKYDMXX_")) %>%
  dplyr::filter(!duplicated(sample_id)) %>%
  magrittr::set_rownames(.$sample_id)

saveRDS(ps.merged, "output/rds/ps_merged.rds")

## Train model again on full mock dataset and predict unknowns
final_logistic_glm <- glmnet_wf  %>%
  fit(mock_dat)

final_predictions_glm <- final_logistic_glm %>%
  predict(new_data = model_df, type="prob") %>%
  bind_cols(model_df) %>%
  mutate(outcome = factor(outcome, levels = c("1", "0"))) %>%
  mutate(.pred = factor(make_two_class_pred(.pred_1, levels(outcome), threshold = .5)))%>%
  mutate(diff = !.pred == outcome)


# Get importance of predictors
final_logistic_glm %>%
  pull_workflow_fit() %>% 
  vip::vi() %>%
  unnest() %>%
  group_by(Sign) %>%
  ggplot(aes(x = Importance, y = Variable, fill = Sign)) +
  geom_col() +
  labs(y = NULL)

taxon_roc <- final_predictions_glm %>%
  filter(pcr_primers =="fwhF2-fwhR2n") %>%
  mutate(outcome = factor(outcome, levels = c("0", "1"))) %>% 
  filter(!is.na(outcome)) %>%
  filter(taxon %in% exp$taxon) %>%
  group_by(taxon) %>%
  group_modify(~{
    # Add 1 fake FP observation per taxon to allow ROC curve
    if(length(unique(.x$outcome)) == 1){
      fake_obs <- .x %>%
        dplyr::slice(1) %>%
        mutate(.pred_0 = 0.9999, .pred_1 = 0.0001, outcome=as.factor("0"))
      .x <- .x %>%
        bind_rows(fake_obs)  
    }
    return(.x)
  })

gg.roc_curve <- taxon_roc %>%  
  roc_curve(truth = outcome, .pred_0) %>%
  left_join(taxon_roc %>%
  roc_auc(truth = outcome, .pred_0)) %>%
   mutate(taxon = taxon %>%
           str_replace("_", " ") ) %>%
  #filter(.threshold < Inf & .threshold > -Inf) %>%
  filter(taxon %in% c("Drosophila suzukii",
                      "Drosophila subpulchrella",
                      "Drosophila biarmipes",
                      "Drosophila immigrans",
                      "Drosophila hydei",
                      "Drosophila melanogaster",
                      "Drosophila simulans",
                      "Scaptodrosophila lattivitata"
                      )) %>%
  mutate(taxon = paste0(taxon, "\n(AUC: ", round(.estimate, 3), ")")) %>%
  ggplot(aes(x = 1 - specificity, y = sensitivity, colour= taxon)) +
  geom_path(size=1, position=position_dodge(width=0.03), alpha=0.8) +
  geom_abline(lty = 3) + 
  facet_wrap(~taxon, ncol = 4) +
  coord_equal() +
  scale_color_brewer(palette="Paired")+
  scale_x_continuous(labels = scales::percent) +
  scale_y_continuous(labels = scales::percent) +
  base_theme+
  theme(legend.text = element_text(face="italic"),
        legend.position = "none")+
  labs(colour = "Taxon",
       x = "1 - Specificity",
       y = "Sensitivity")

gg.roc_curve

pdf(file="fig/supplementary/roc_curve.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg.roc_curve)
try(dev.off(), silent=TRUE)


final_predictions_glm %>%
    filter(pcr_primers =="fwhF2-fwhR2n") %>%
  mutate(taxon = taxon %>% str_replace("_", " ")) %>%
  filter(taxon %in% c("Drosophila suzukii",
                             "Drosophila subpulchrella",
                             "Drosophila biarmipes"),outcome == 1 
         ) %>%
  dplyr::select(-where(~is.list(.x))) %>%
  mutate(total_seq = log10(total_seq)) %>%
  pivot_longer(cols=c("total_abund", "total_seq", "expected_RA"),
               names_to = "metric",
               values_to = "value") %>%
  ggplot(aes(x = value, y=.pred_1, colour=taxon, group=taxon))+
  geom_point(alpha=0.5) +
  geom_smooth(method="lm")+
  coord_cartesian(ylim=c(0,1))+
  facet_wrap(metric~., scales="free_x", ncol=1) +
  base_theme+
  scale_colour_brewer(palette="Paired")+
  theme(legend.position="bottom")


# Make a probability surface
newdat <- tidyr::crossing(
  abundance_RA= seq(0, 0.0001,len=1000),
  n_rep = seq(1, 6,1),
  n_extract = seq(1, 2, 1),
  fcid = unique(final_predictions_glm$fcid),
  pcr_primers = unique(final_predictions_glm$pcr_primers),
  )

gg.prob_surface <- bind_cols(newdat, final_logistic_glm %>%
  predict(new_data = newdat, type="prob")) %>%
  mutate(.pred = factor(make_two_class_pred(.pred_1, c("0", "1"), threshold = .5)))%>%
  filter(fcid == "HLVKYDMXX") %>%
  pivot_longer(starts_with("n_"),
               names_to="rep_type",
               values_to="rep_n") %>%
  mutate(rep_type = rep_type %>% str_replace("n_extract", "Extraction") %>% str_replace("n_rep", "PCR")) %>%
  ggplot(aes(x = abundance_RA, y = rep_n))+
  geom_contour_filled(aes(z = .pred_1, colour=.pred_1), alpha=0.8)+
  scale_colour_viridis_d() +
  #facet_grid(~fcid)+
  new_scale_colour()+
  new_scale_fill() +
  geom_point(data=final_predictions_glm  %>%
  pivot_longer(starts_with("n_"),
               names_to="rep_type",
               values_to="rep_n") %>%
      mutate(rep_type = rep_type %>% str_replace("n_extract", "Extraction") %>% str_replace("n_rep", "PCR")) %>%
               filter(fcid == "HLVKYDMXX", abundance_RA < 0.001, rep_n > 0), aes(colour= .pred, fill=.pred_1),
             position = position_jitter(width=0, height=0.01), alpha=1,shape=21, show.legend=FALSE ) +
  scale_colour_manual(values=c('0'="#d7191c", '1'="#2b83ba"))+
  scale_fill_viridis_c() +
  scale_x_continuous(labels = scales::percent, expand = c(0,0)) +
  base_theme +
  facet_grid(rep_type~., scales = "free")+
  coord_cartesian(xlim=c(0,0.0001), expand = c(0,0))+
  theme(legend.position = "right") +
  labs(x = "Relative abundance",
       y = "Number of replicates") 


gg.prob_surface

# Calculate metrics
target_det <- final_predictions_glm %>%
    filter(fcid == "HLVKYDMXX") %>%
    filter(pcr_primers =="fwhF2-fwhR2n") %>%
  mutate(taxon = taxon %>%
           str_replace("_", " ")) %>%
  filter(taxon %in% c("Drosophila suzukii",
                             "Drosophila subpulchrella",
                             "Drosophila biarmipes")) 

#Problematic detections check
problem <- target_det %>% 
  filter(abundance  > 0) %>%
  filter(.pred == 1 & is.na(outcome))

#Disagreements with mock
diss <- final_predictions_glm %>% 
  mutate(diff = !outcome  == .pred) %>%
  filter(diff)

# Number of samples
target_det %>% 
  filter(abundance  >0) %>%
  mutate(success = case_when(
    outcome ==1 & .pred == 1 ~ "TP",
    outcome == 1 & .pred == 0 ~ "FN",
    outcome == 0 & .pred == 1 ~ "FP",
    outcome == 0 & .pred == 0 ~ "TN"
  )) %>%
  group_by(taxon) %>%
  summarise(TP = sum(success=="TP", na.rm = TRUE), FN = sum(success=="FN", na.rm = TRUE), 
            FP= sum(success == "FP", na.rm = TRUE), TN = sum(success == "TN", na.rm = TRUE),
            total_pos = sum(success %in% c("TP", "FN")))

multimetric <- metric_set(accuracy, bal_accuracy, sens, yardstick::spec, precision, recall, ppv, npv)
final_metrics <- target_det %>%
  group_by(taxon) %>%
  multimetric(truth = as.factor(outcome), estimate = .pred)

final_metrics

# Dropouts as function of community size 
group.colors <- c(`True Positive` = "#abdda4", `False Positive` = "#d7191c", `False Negative` ="#2b83ba", `True Negative` ="#cccccc")

gg.sens <- target_det %>% 
  mutate(success = case_when(
    outcome ==1 & .pred == 1 ~ "TP",
    outcome == 1 & .pred == 0 ~ "FN",
    outcome == 0 & .pred == 1 ~ "FP",
    outcome == 0 & .pred == 0 ~ "TN"
  )) %>%
  mutate(expected_RA = case_when(
    is.na(expected_RA) ~ as.numeric(0),
    TRUE ~ as.numeric(expected_RA)
  )) %>% 
  #mutate(abundance_RA = case_when(
  #  abundance_RA==0 ~ as.numeric(0.0001),
  #  TRUE ~ as.numeric(abundance_RA)
  #)) %>% 
  mutate(community = case_when(
    str_detect(sample_id, "_D[0-9]|_DM|_DL") ~ "Mock",
    TRUE ~ "Trap"
  )) %>% 
  dplyr::filter(success %in% c("TP", "FN", "FP")) %>%
    left_join(final_metrics  %>% 
              filter(.metric %in% c("sens", "spec", "ppv", "npv", "bal_accuracy")) %>%
              mutate(.estimate = scales::percent(.estimate, accuracy=0.1)) %>%
              pivot_wider(names_from=.metric,
                          values_from=.estimate)) %>%
  mutate(taxon = taxon %>%
         factor(levels = c("Drosophila suzukii",
                             "Drosophila subpulchrella",
                             "Drosophila biarmipes"))) %>%
  mutate(success = case_when(
    success == "TP" ~ "True Positive",
    success == "FP" ~ "False Positive",
    success == "FN" ~ "False Negative"
  )) %>%
  ggplot(aes(x = expected_RA, y=abundance_RA, fill=success, shape=community, size=n_rep)) + 
  geom_point(size=2, alpha=1, position=position_jitter(width=.001, height=0.001), shape=21, colour="black") + 
  geom_abline(lty = 2, colour = "gray80", size = 1) +
  facet_wrap(taxon~.) +
  base_theme +
  theme(legend.position = "none",
        strip.text.x = element_text(face="italic"))+
  #scale_colour_viridis_d()+
  scale_fill_manual(values = group.colors) +
  coord_cartesian(clip = "off") +
  #annotation_logticks(sides="lb", outside = TRUE) +
  scale_x_continuous(labels = scales::percent, trans = pseudo_log_trans(0.0001),
                     breaks=c(0, 0.00025,0.001, 0.005, 0.01, 0.05, 0.1))+
  scale_y_continuous(labels = scales::percent, trans = pseudo_log_trans(0.0001),
                     breaks=c(0, 0.00025,0.001, 0.005, 0.01, 0.05, 0.1))+
  labs(x = "Expected relative abundance (from specimens)",
       y = "Observed relative abundance (from sequencing)",
       shape = "Community Type",
       colour = "Detection Success")

gg.sens

# Filter predicted switching
keeptab <- final_predictions_glm %>%
  dplyr::select(.pred, sample_id , otu) %>%
  mutate(keep = case_when(
    .pred == 0 ~ 0,
    .pred == 1 ~ 1
  )) %>%
  dplyr::select(-.pred) %>%
  pivot_wider(names_from = otu,
              values_from = keep,
              values_fill = 0) %>%
  dplyr::slice(match(rownames(otu_table(ps.merged)), sample_id)) %>%
  column_to_rownames("sample_id")%>%
  dplyr::select(colnames(otu_table(ps.merged))) %>%
  as.matrix()


#Check for disagreements
colnames(otu_table(ps.merged))[!colnames(otu_table(ps.merged)) %in% colnames(keeptab)]
rownames(otu_table(ps.merged))[!rownames(otu_table(ps.merged)) %in% rownames(keeptab)]

#get current OTU table
otutab <- otu_table(ps.merged) %>%
  as("matrix") 

#Replace taxa predicted as False positives
ps2 <- ps.merged
otu_table(ps2) <- otu_table(otutab * keeptab, taxa_are_rows = FALSE)

# Check staxa that were dropped
dropped <- taxa_sums(ps2) %>%
  as_tibble(rownames = "OTU") %>%
  left_join(taxa_sums(ps1) %>%
  as_tibble(rownames = "OTU") %>%
    dplyr::rename(pre_filt = value)) %>%
  filter(value == 0) 

# Remove zero taxa
ps2 <- filter_taxa(ps2, function(x) mean(x) > 0, TRUE) #Drop missing taxa from table
message(nsamples(ps_model) - nsamples(ps2), " Samples and ", ntaxa(ps_model) - ntaxa(ps2), " taxa under read threshold Dropped")

## Write out final phyloseq
saveRDS(ps2, "output/rds/ps_filtered.rds")

## Write out filtered summaries
dir.create("output/csv/filtered")
speedyseq::psmelt(ps2) %>%
  filter(Abundance > 0) %>%
  write_csv("output/csv/filtered/filtered_data.csv")

#Summary export
seqateurs::summarise_taxa(ps2, "Species", "sample_name") %>%
  filter(!str_detect(sample_name, "NTC")) %>%
  spread(key="sample_name", value="totalRA") %>%
  write.csv(file = "output/csv/filtered/spp_sum.csv")

seqateurs::summarise_taxa(ps2, "Genus", "sample_name") %>%
  spread(key="sample_name", value="totalRA") %>%
  write.csv(file = "output/csv/filtered/gen_sum.csv")
```

# Detection / Switching removal figure

```{r Detection}
pred_boots <- readRDS(file="output/index_switch_bootstraps.rds")

group.colors <- c(`True Positive` = "#abdda4", `False Positive` = "#d7191c", `False Negative` ="#2b83ba", `True Negative` ="#cccccc")

gg.switch_resolved <- pred_boots %>%
  filter(!est_type == "pred_rf") %>%
  #filter(fcid=="HLVKYDMXX") %>%
  group_by(est_type, outcome_post, fcid) %>%
  summarise(count = n(), count_fp = sum(outcome_post == "FP")) %>%
  group_by(est_type, fcid) %>%
  mutate_at(vars(count), ~ . / sum(.) ) %>%
  filter(!(fcid=="CB3DR" & est_type=="pred_pos")) %>%
  mutate(est_type = est_type %>%
           str_replace("pred_logit$", "Logistic Reg") %>%
           str_replace("pred_logit_reps$", "Logistic Reg + Reps") %>%
           str_replace("pred_mock", "Mock Communities") %>%
           str_replace("pred_naive", "0.01% Threshold") %>%
           str_replace("pred_pos", "Positive control") %>%
           str_replace("pred_uncorrected", "Uncorrected") %>%
           str_replace("pred_index", "Unassigned indices")
         ) %>%
  mutate(fcid = case_when(
    fcid == "HLVKYDMXX" ~ "NovaSeq",
    TRUE ~ "MiSeq",
  )) %>%
  mutate(est_type = factor(est_type,
                           levels= c(
    "Logistic Reg + Reps",
    "Logistic Reg",
    "Mock Communities",
    "Positive control",
    "Unassigned indices",
    "0.01% Threshold",
    "Uncorrected"))) %>%
  ggplot(aes(x=est_type, y=count, fill=outcome_post)) + 
  geom_col(position="stack", alpha = 0.8) +
  base_theme+
  coord_flip()+
  facet_wrap(~fcid, ncol=2) +
  scale_y_continuous(labels = scales::percent)+ 
  scale_fill_manual(values=group.colors) + 
  theme(legend.position = "bottom",
        axis.ticks = element_line()) +
  labs(y = NULL,
       x = NULL,
       fill = "Detection Success")

# Create multifig
#gg.detection_multi <- gg.switch_resolved - ( gg.prob_surface / gg.sens ) + plot_annotation(tag_levels = #"A") + plot_layout(widths = c(1,2))
#
#gg.detection_multi
#
# Multifig2
gg.detection_multi <- gg.switch_resolved / gg.sens + plot_annotation(tag_levels = "A") 

gg.detection_multi


pdf(file="fig/Fig3_detection.pdf", width = 10, height = 8 , paper="a4r")
  plot(gg.detection_multi)
try(dev.off(), silent=TRUE)

```

# Primer comparison

## Overview of run 1

```{r run 1 overview}
# Subset to mocks
ps_run1 <- ps1

tax_table(ps_run1)[,8] <- tax_table(ps_run1)[,8]%>%
  str_replace("Drosophila albomicans", "Drosophila immigrans") %>%
  str_replace("Drosophila nasuta", "Drosophila immigrans") %>%
  str_replace("Drosophila hypocausta", "Drosophila immigrans") %>%
  str_replace("Drosophila pulaua", "Drosophila immigrans") %>%
  str_replace("Drosophila kohkoa", "Drosophila immigrans") %>%
  str_replace("Drosophila rubida", "Drosophila immigrans") %>%
  str_replace("Drosophila sulfurigaster", "Drosophila immigrans") %>%
  str_replace(" ", "_")

ps_run1 <- ps_run1 %>%
  speedyseq::tax_glom(taxrank = "Species") %>%
  subset_samples(fcid %in% c("CB3DR")) %>%
  subset_samples(!str_detect(sample_names(ps_run1),"blank")) %>%
  filter_taxa( function(x) mean(x) > 0, TRUE) 

ps_run1 %>% 
  speedyseq::psmelt() %>%
  group_by(sample_id) %>%
  mutate_at(vars(Abundance), ~ . / sum(.) ) %>% #Convert to proportions
  mutate(plotlabel = case_when(
    Abundance >= 0.01  ~ Species, # Change this to whatever taxrank we want
    Abundance < 0.01 ~ as.character(NA)
    )) %>%
  ggplot(aes(x=sample_name, y=Abundance, fill=plotlabel)) +
  geom_col(position="stack") + 
  facet_grid(pcr_primers~type, scales="free") +
  scale_y_continuous(labels = scales::percent) +
  coord_flip()+
  base_theme+
  theme(legend.position = "bottom") +
  labs(x = "Sample Name",
       y= "Relative abundance",
       fill="Species",
       title = "Run 1 - Primer testing")

# Reads per sample
ps_run1 %>% 
  speedyseq::psmelt() %>%
  group_by(sample_id, sample_name, pcr_primers) %>%
  summarise(Abundance = sum(Abundance)) %>%
  ggplot(aes(x = sample_name, y = Abundance))+
  geom_col() +
  geom_text(aes(label=Abundance),angle=90,hjust=1,vjust=0.5, colour="white")+
  facet_grid(~pcr_primers,scales="free_x", drop=TRUE) + 
  base_theme

# Run 1 summary stats
ps_run1 %>% 
  speedyseq::psmelt() %>%
  group_by(sample_id, sample_name, pcr_primers) %>%
  summarise(Abundance = sum(Abundance)) %>%
  ungroup() %>%
  summarise(total = sum(Abundance), mean=mean(Abundance), sd = sd(Abundance),
            se =  sd(Abundance)/sqrt(length(Abundance)),upper = range(Abundance)[2], lower=range(Abundance)[1])

# Run 2 summary stats
ps1 %>% 
  speedyseq::psmelt() %>%
  filter(fcid == "HLVKYDMXX")%>% 
  group_by(sample_id, sample_name, pcr_primers) %>%
  summarise(Abundance = sum(Abundance)) %>%
  ungroup() %>%
  summarise(total = sum(Abundance), mean=mean(Abundance), sd = sd(Abundance),
            se =  sd(Abundance)/sqrt(length(Abundance)),upper = range(Abundance)[2], lower=range(Abundance)[1])

# Taxa per primer
ps_run1 %>% 
  speedyseq::psmelt() %>%
  filter(Abundance > 0, !is.na(Abundance)) %>%
  mutate(type = case_when(
    str_detect(sample_name,"D100M") ~ "MOCK",
    TRUE ~ "TRAP"
  ))%>%
  group_by(type, pcr_primers) %>%
  summarise(unique = n_distinct(Species)) 

# Heatmap separate for primers
heatmap_dat <-  ps_run1 %>%
  speedyseq::psmelt() %>%
  dplyr::select(sample_id, OTU, sample_name, Abundance, pcr_primers, Species, type) %>%
  filter(!str_detect(Species, "__")) %>%
  mutate(sample_name = sample_name %>% 
           str_remove("CB3DR_") %>%
          str_remove_all("BF1-BR1-|SauronS878-HexCOIR4-|fwhF2-HexCOIR4-|fwhF2-fwhR2n-"),
         Species = str_replace(Species, "_", " "),
         type = case_when(
           str_detect(sample_name, "D100") ~ "Mock",
           TRUE ~ "Trap"
         )) %>%
  group_by(sample_id, pcr_primers) %>%
  mutate_at(vars(Abundance), ~ . / sum(.) ) %>%
  ungroup() %>%
  filter(Abundance > 0) %>% 
  mutate(Abundance = case_when(
    Abundance <0.00011 ~ as.numeric(NA),
    Abundance > 0.00011 ~ Abundance
  )) %>%
  mutate(Abundance = na_if(Abundance, 0))  %>%
  left_join(read_csv("sample_data/expected_quant_merged.csv") %>%
  dplyr::rename(sample_name = Sample) %>%
  pivot_longer(-sample_name,
               names_to= "Species",
               values_to= "expected") %>%
    mutate(Species = Species %>% str_replace("_", " ")), by = c("sample_name","Species")) %>%
  mutate(expected = na_if(expected, 0))  %>%
  mutate(outcome = case_when(
    type == "Mock" & Abundance > 0 & is.na(expected) ~ "FP",
    type == "Mock" & is.na(Abundance) & expected > 0 ~ "FN"
  )) %>%
  filter(!(is.na(Abundance) & is.na(outcome)))

#Taxa that were different across primers
heatmap_dat %>%
  filter(Abundance > 0) %>% # This is ruin
  group_by(Species) %>%
  mutate(diff = n_distinct(pcr_primers)) %>%
  ungroup() %>%
  filter(diff < 4) %>%
  dplyr::select(sample_name, pcr_primers, Species, diff) %>%
  distinct()

# False positives per primer
heatmap_dat %>%
  filter(!is.na(outcome)) %>%
  group_by(outcome, pcr_primers) %>%
  summarise(unique = n_distinct(Species)) 


# Plot tree
keeptips <- heatmap_dat %>% 
  filter(Abundance > 0) %>%
  dplyr::select(OTU, Species) %>%
  distinct()

tree <- phy_tree(ps_run1)
tree <- drop.tip(tree, tree$tip.label[!tree$tip.label %in% keeptips$OTU])

tree$tip.label <- tree$tip.label %>%
  tibble::enframe(name=NULL, value="OTU") %>%
  left_join(keeptips) %>%
  pull(Species)

mock_tree <- ggtree(tree, ladderize = TRUE) + 
  geom_tiplab(align=TRUE)
gg.run1_heatmap <- heatmap_dat %>%
  left_join(mock_tree$data %>% dplyr::rename(Species = label)) %>%
  mutate(Species = factor(Species),
         sample_name = factor(sample_name, levels= c("D100M1", "D100M2", "D100M3", "D100M4", "D100M5", "DM0-FF", "DM0Sach","DM6-SPD", "DM10-FF", "DM10-SPD"))) %>%
    ggplot(aes(x=sample_name, y=fct_reorder(Species, y), fill=Abundance)) +
    geom_tile() +
  scale_fill_viridis_c(labels = scales::percent, na.value = NA, alpha=0.9) +
  new_scale_fill() +
  geom_tile(aes(fill = outcome))+
  scale_fill_manual(values=c("FP" = "#e41a1c", "FN"= "grey60"), na.value = NA) +
  base_theme+
  theme(legend.position = "right",
        axis.text.y = element_text(face="italic"),
        axis.title.y = element_blank())+
    labs(x="Sample",
         y="Taxon") +
  facet_grid(~pcr_primers, drop=TRUE)

mock_tree + gg.run1_heatmap + plot_layout(widths = c(1,4))
```

## Bias

```{r Drosophila bias}
ps_bias <- ps_run1

exp <- read_csv("sample_data/expected_quant_merged.csv") %>%
  dplyr::rename(sample_name = Sample) %>%
  pivot_longer(-sample_name,
               names_to= "taxon",
               values_to= "expected") %>%
  mutate(taxon = str_replace(taxon, pattern=" ",replacement="_"),
         sample_name = str_remove(sample_name, "-exp*.$")) %>%
  dplyr::filter(!is.na(sample_name),
  !str_detect(sample_name, "CM[0-9]|CT[0-9]|CML[0-9]|Syn"),
  expected > 0) %>%
  distinct()

#plot expected
exp %>% 
  dplyr::filter(str_detect(sample_name, "^D")) %>%
  group_by(sample_name) %>%
  mutate_at(vars(expected), ~ . / sum(.) ) %>% #Convert to proportions
  ggplot(aes(x=sample_name, y=expected, fill=taxon)) +
  geom_col(position="stack") + 
  scale_fill_brewer(palette="Spectral") +
  scale_y_continuous(labels=scales::percent) +
  base_theme+
  theme(legend.position = "bottom",
        axis.text.x = element_text(angle=45, hjust=1)) +
  labs(x = "Sample Name",
       y= "Relative abundance",
       title="Expected mock communities",
       fill="Species")

#Get observed
sam <- speedyseq::psmelt(ps_bias) %>%
  janitor::clean_names() %>%
  mutate(taxon = species) %>%
  filter(!str_detect(genus, "__")) %>%#Remove unclassified
  dplyr::select(sample_name, taxon, abundance, pcr_primers, fcid, material_type = type) %>%
  mutate(sample_name = sample_name %>%
           str_remove_all("BF1-BR1-|SauronS878-HexCOIR4-|fwhF2-HexCOIR4-|fwhF2-fwhR2n-") %>%
           str_remove("HLVKYDMXX_"),
         sample_id = paste0(fcid, "_", sample_name),
         sample_name = sample_name %>%
           str_remove_all("HLVKYDMXX_|CK3HD_|CB3DR_|CJKFJ_")) 


# create a good pseudocount
# this was chosen to give a proportion that is less than 1 divided by the maximum sample read depth
# and so is expected to have negligible effect on the resulting estimates.
pseudocount <- sam %>% 
  group_by(sample_id) %>%
  summarise(sum = sum(abundance)) %>%
  pull(sum)%>% 
  max() %>% 
  {1 / .}

#Join expected and observed tables 
joint <- sam %>%
  filter(taxon %in% exp$taxon,
         sample_name %in% exp$sample_name) %>%
  left_join(exp, by = c("sample_name","taxon")) %>%
  dplyr::rename(observed = abundance) %>%
  filter(observed > 0) %>%
  mutate(expected = replace_na(expected, 0),
         observed0 = (observed + pseudocount) * (expected > 0),
         error = observed0 / expected) %>%
  distinct() %>%
  group_by(sample_id, pcr_primers) %>%
  mutate_at(vars(observed, expected), ~ . / sum(.) ) %>% #Convert to proportions
  filter(observed > 0.00011) #Remove taxa under threshold

# Visualise the error in all pairwise ratios
gg.ratio <- joint %>%
  dplyr::filter(expected > 0 ) %>%
  dplyr::mutate(Taxon = taxon %>%
                 str_replace("^.+_", paste0(str_extract(taxon, "."), "_"))) %>%
    compute_ratios(group_vars = c("sample_id", "material_type", "pcr_primers")) %>%
  mutate(Pair = paste(Taxon.x, Taxon.y, sep = ":")) %>% 
  ggplot(aes(Pair, expected, colour=sample_id)) +
  geom_hline(yintercept = 1, alpha=0.8) +
  geom_jitter(alpha=0.7) +
  scale_y_log10() +
  base_theme+
  facet_grid(pcr_primers~.)+
    theme(legend.position = "none",
          axis.text.x = element_text(angle=45, hjust=1))+
  labs(y= "Error in taxon ratios (log10)") 

gg.ratio


# Fit the metacal model to each pcr primer set and community type
fit_metacal <- joint %>%
  filter(material_type=="DrosMock") %>%
  filter(expected > 0, observed > 0) %>%
  group_by(pcr_primers, material_type)  %>%
  nest() %>%
    mutate(
        fit = purrr::map(data, function(x){
        obs_mat <- x %>%
          dplyr::select(sample_id, taxon, observed)%>%
          pivot_wider(id_cols= sample_id,
                      names_from = "taxon",
                      values_from = "observed",
                      values_fill = list(observed=0))%>%    
          column_to_rownames("sample_id") %>%
          as.matrix() 
        
        exp_mat <- x %>%
          dplyr::select(sample_id, taxon, expected)%>%
          pivot_wider(id_cols= sample_id,
                      names_from = "taxon",
                      values_from = "expected",
                      values_fill = list(expected=0))%>%    
          column_to_rownames("sample_id") %>%
          as.matrix() 
                
       fit_mc <- estimate_bias(
          obs_mat,
          exp_mat, 
          margin = 1, # samples as rows
          boot = TRUE
        ) %>%
         summary()
       out <- fit_mc$coefficients
       return(out)
        })) %>%
    unnest(fit) %>%
  mutate(lower = estimate / gm_se^2, upper = estimate * gm_se^2)


# See how well bias estimates fit the data
preds <- joint %>%
  left_join(fit_metacal) %>%
  dplyr::mutate(predicted = expected * estimate) %>%
  group_by(sample_id, pcr_primers) %>%
  mutate_at(vars(predicted), ~ . / sum(., na.rm=TRUE) ) %>%
  filter(expected > 0) %>%
  filter(observed > 0)

# Get RMSE
errors <- preds %>%
  group_by(pcr_primers) %>%
  rmse(truth = observed, estimate = predicted) %>%
  dplyr::select(pcr_primers, rmse = .estimate)

gg.bias_fits <- preds %>%
  mutate(taxon = str_replace(taxon, "_", " ")) %>%
  left_join(errors) %>%
  mutate(rmse = paste0("RMSE = ", scales::percent(rmse, 2))) %>%
  ggplot(aes(x=predicted, y=observed, color = taxon))+
  geom_abline(intercept = 0, slope = 1, color = "grey") +
    geom_jitter(width = 0.1, height = 0, alpha=0.7, size=2) +
  geom_text(aes(x=0.05, y=0.9, label=rmse), check_overlap = TRUE, inherit.aes = FALSE)+
  facet_wrap(~pcr_primers)+
  coord_fixed()+
  scale_color_brewer(palette = "Paired")+
  scale_x_continuous(trans = scales::pseudo_log_trans(1e-2), labels=scales::percent, breaks=c(0, 0.025, 0.05, 0.1, 0.25, 0.5, 1), limits=c(0,1))+
  scale_y_continuous(trans = scales::pseudo_log_trans(1e-2), labels=scales::percent, breaks=c(0, 0.025, 0.05, 0.1, 0.25, 0.5, 1), limits=c(0,1))+
  base_theme +
  theme(
        panel.spacing.x = unit(1, "lines"),
        legend.position = "right",
        legend.text = element_text(face = "italic", size=10)
    ) +
   labs(x = "log-odds(Predicted proportion)", 
     y = "log-odds(Observed proportion)",
     colour = "Taxon") 


gg.bias_fits

# Write out bias explained for supplementary 
pdf(file="fig/supplementary/bias_model_fit.pdf", width = 8, height = 6 , paper="a4r")
  plot(gg.bias_fits)
try(dev.off(), silent=TRUE)

# Plot bias estimates
gg.bias <- preds %>%
  ungroup() %>%
  dplyr::select(taxon, estimate, upper, lower, pcr_primers) %>%
  distinct()%>%
  left_join(mock_tree$data %>% dplyr::mutate(taxon = label %>% str_replace(" ", "_")))%>%
  filter(!is.na(estimate)) %>%
  mutate(taxon = taxon %>% 
           str_replace("^Drosophila_", "D. ")%>%
           str_replace("^Scaptodrosophila_", "Sca. ")) %>%
  mutate(taxon = factor(taxon))%>%
  ggplot(aes(x = fct_reorder(taxon, y), y = estimate, colour = pcr_primers, group = pcr_primers)) +
  geom_hline(yintercept = 1, alpha=0.5, colour = "grey20") +
  geom_pointrange(aes(ymin = lower, ymax = upper), position=position_dodge(width=0.6))+
  scale_y_log10()+
  base_theme+
  scale_colour_brewer(palette="Paired") +
  labs(
    x = NULL,
    colour="PCR primers",
    y = "Efficiency / Geometric mean") +
  theme(axis.text.x = element_text(face = "italic"),
        legend.position = "bottom")

gg.bias

#bias is missing for BF1 biarmipes because it was under filtering threshold
```

## Multifig1

Comparison of primers for bias, off-target amp, and PCA of overlap

```{r run 1 multifig}
fig1a <- mock_tree + gg.run1_heatmap + plot_layout(widths = c(1,6))

fig1 <- fig1a / gg.bias + plot_layout(heights = c(2,1))+ plot_annotation(tag_levels = 'A') & theme(plot.margin = unit(c(1,1,1,1), "mm"))

fig1

# Write out figure 1
pdf(file="fig/Fig1_primer_comparison.pdf", width = 11, height = 8 , paper="a4r")
  plot(fig1)
try(dev.off(), silent=TRUE)
```

# Supplementary Comparison of primer replicates for detection and bias

Run 2 - see the differences in bias between replicates

```{R run2}
# RUN 2
ps_run2 <- ps1 %>%
  subset_samples(fcid %in% c("CK3HD")) %>%
  filter_taxa( function(x) mean(x) > 0, TRUE) %>%
  speedyseq::tax_glom(taxrank="Species")

exp <- read_csv("sample_data/expected_quant_merged.csv") %>%
  dplyr::rename(sample_name = Sample) %>%
  pivot_longer(-sample_name,
               names_to= "taxon",
               values_to= "expected") %>%
  mutate(taxon = str_replace(taxon, pattern="_",replacement=" "),
         sample_name = str_remove(sample_name, "-exp*.$")) %>%
  dplyr::filter(!is.na(sample_name),
  !str_detect(sample_name, "CM[0-9]|CT[0-9]|CML[0-9]|Syn"),
  expected > 0) %>%
  distinct()

#Get observed
sam <- ps_run2 %>%
  speedyseq::psmelt() %>%
  janitor::clean_names() %>%
  mutate(taxon = species) %>%
  filter(!str_detect(genus, "__")) %>%#Remove unclassified
  dplyr::select(sample_name, sample_id, amp_rep, taxon, abundance, pcr_primers, fcid, material_type = type) %>%
  mutate(sample_name = sample_name %>%
           str_remove_all("BF1-BR1-|SauronS878-HexCOIR4-|fwhF2-HexCOIR4-|fwhF2-fwhR2n-") %>%
           str_remove_all("HLVKYDMXX_|CK3HD_|CB3DR_|CJKFJ_"))


# create a good pseudocount
# this was chosen to give a proportion that is less than 1 divided by the maximum sample read depth
# and so is expected to have negligible effect on the resulting estimates.
pseudocount <- sam %>% 
  group_by(sample_id) %>%
  summarise(sum = sum(abundance)) %>%
  pull(sum)%>% 
  max() %>% 
  {1 / .}

#Join tables 
joint <- sam %>%
  filter(taxon %in% exp$taxon,
         sample_name %in% exp$sample_name,
         abundance > 0) %>%
  left_join(exp, by = c("sample_name","taxon")) %>%
  dplyr::rename(observed = abundance) %>%
  mutate(expected = replace_na(expected, 0),
         observed0 = (observed + 0.5) * (expected > 0),
         error = observed0 / expected) %>%
  distinct() %>%
  group_by(sample_id) %>%
  mutate_at(vars(observed, expected), ~ . / sum(.) ) %>% #Convert to proportions
  ungroup() 

# Fit the metacal model to each pcr primer set and amplification rep
fit_metacal <- joint %>%
  filter(material_type=="DrosMock") %>%
  filter(expected > 0) %>%
  group_by(pcr_primers, amp_rep) %>%
  nest() %>%
    mutate(
        fit = purrr::map(data, function(x){
        obs_mat <- x %>%
          dplyr::select(sample_id, taxon, observed)%>%
          pivot_wider(id_cols= sample_id,
                      names_from = "taxon",
                      values_from = "observed",
                      values_fill = list(observed=0))%>%    
          column_to_rownames("sample_id") %>%
          as.matrix() 
        
        exp_mat <- x %>%
          dplyr::select(sample_id, taxon, expected)%>%
          pivot_wider(id_cols= sample_id,
                      names_from = "taxon",
                      values_from = "expected",
                      values_fill = list(expected=0))%>%    
          column_to_rownames("sample_id") %>%
          as.matrix() 
                
       fit_mc <- estimate_bias(
          obs_mat,
          exp_mat, 
          margin = 1, # samples as rows
          boot = TRUE
        ) %>%
         summary()
       out <- fit_mc$coefficients
       return(out)
        })) %>%
    unnest(fit) %>%
  mutate(lower = estimate / gm_se^2, upper = estimate * gm_se^2)


# See how well bias estiamte fits data
preds <- joint %>%
  left_join(fit_metacal) %>%
  dplyr::mutate(predicted = expected * estimate) %>%
  group_by(sample_id) %>%
  mutate_at(vars(predicted), ~ . / sum(., na.rm=TRUE) ) %>%
  filter(expected > 0) %>%
  filter(observed > 0)

# Plot bias estimates for the 3 separate primers
gg.biasreps <- preds %>%
  filter(pcr_primers == "fwhF2-fwhR2n")%>%
  ungroup() %>%
  dplyr::select(taxon, estimate, upper, lower, pcr_primers, amp_rep) %>%
  distinct()%>%
  mutate(amp_rep = factor(amp_rep, levels = c("1","2","3"))) %>%
  mutate(taxon = str_replace(taxon, "^.*_", "D. ")) %>%
  ggplot(aes(x = taxon, y = estimate, colour = amp_rep, group =amp_rep)) +
  geom_hline(yintercept = 1, alpha=0.5, colour = "grey20") +
  geom_pointrange(aes(ymin = lower, ymax = upper), position=position_dodge(width=0.5))+
  #facet_grid(~pcr_primers) +
  coord_flip() +
  base_theme+
  scale_y_log10()+
  scale_colour_brewer(palette="Paired") +
  labs(
    x = NULL,
    y = "Efficiency / Geometric mean",
    colour = "Tagged PCR primer") +
  theme(axis.text.y = element_text(face = "italic"),
        legend.position = "bottom")

gg.biasreps

# Write supplementary figure
pdf(file="fig/supplementary/run2_biasreps.pdf", width = 6, height = 4 , paper="a4r")
  plot(gg.biasreps)
try(dev.off(), silent=TRUE)
```

# Trap catch & Diversity

## Insects trapped

```{r Trapped insects}
trapdat <- read_csv("Fieldwork/trap_data_clean.csv") %>%
  janitor::clean_names()%>%
  mutate(catch = as.numeric(catch)) %>%
  mutate(orchard_type = case_when(
         str_detect(fruit, "Peach") ~ "Stonefruit",
         str_detect(fruit, "Nectarine") ~ "Stonefruit",
         str_detect(fruit, "Stonefruit") ~ "Stonefruit",
         str_detect(fruit, "Cherries") ~ "Cherry",
         TRUE ~ "Cherry"
  )) %>%
  mutate(type = case_when(
    attractant == "Dong Cha" ~ "DC",
    attractant == "ACV" ~ "ACV",
    attractant== "P glycol" ~ "SPD",
    attractant== "FF" ~ "FF"
  )) %>%
  mutate() %>%
  dplyr::filter(!attractant=="Sachet") %>%
  dplyr::select(-treatment, -attractant) %>%
  mutate(type = factor(type, levels =  c("ACV","DC", "SPD", "FF"))) 


# Total catch
trapdat %>%
  group_by(week, type, orchard, orchard_type) %>%
  dplyr::summarise(catch = sum(catch, na.rm=TRUE))
  
# By orchard
trapdat %>%
  group_by(orchard) %>%
  dplyr::summarise(catch = sum(catch, na.rm=TRUE))

# By treatment
trapdat %>%
  group_by(type) %>%
  dplyr::summarise(catch = sum(catch, na.rm=TRUE))

##Total catch by orchard and week
gg.trapweek <- trapdat %>% 
  mutate(week = as.factor(week)) %>%
  ggplot(aes(x=week, y=catch, colour=type, group=week))+
  geom_jitter(size=1, alpha=0.6, width=0.2, height=0.2)+
  geom_boxplot(fill=NA, outlier.colour = NA)+
  facet_grid(type~orchard_type, scales="free") +
  base_theme +
  scale_colour_brewer(palette="Paired") 

##Total catch by orchard and trap type
gg.trapall <- trapdat %>% 
  filter(catch > 0 )%>%
  mutate(metric = "Trapped Individuals") %>%
  #mutate(grouping = paste0(type, "_", orchard_type)) %>%
  ggplot(aes(x=type, y=catch, colour=type, fill=type, group=type))+
  geom_boxplot(width =0.9, alpha=0.7, outlier.colour = NA, colour="black")+
  geom_point(position=position_jitter(width = 0.15, height=0),alpha=0.8, shape=21, colour="black")+
  scale_y_continuous(trans = "log10", breaks=c(1, 5, 20,50, 100, 200, 400))+
  facet_grid(.~orchard_type) +
  base_theme +
  scale_colour_brewer(palette="Paired")  + 
  scale_fill_brewer(palette="Paired")  + 
  labs(x = "Collection method",
       y = "Number Individuals")+
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank()) 


gg.trapall
```

## Number of species

```{r summarise taxa}
# Subset to fwhf2-fwhr2n samples only
ps2 <- readRDS("output/rds/ps_filtered.rds")
ps_run3 <- ps2 %>%
  subset_samples(pcr_primers =="fwhF2-fwhR2n") %>%
  #subset_samples(fcid %in% c("HLVKYDMXX")) %>%
  subset_taxa(Phylum =="Arthropoda") %>%
  #prune_taxa(!str_detect(Species, "Synthetic")) %>%
  filter_taxa( function(x) mean(x) > 0, TRUE)

ps_run3 <-  prune_samples(sample_sums(ps_run3) > 0 , ps_run3)

sample_data(ps_run3)  <- sample_data(ps_run3) %>%
  as("data.frame")%>%
  rownames_to_column("rownames") %>%
  mutate(orchard = case_when(
    str_detect(sample_id, "_M[0-9]") ~ "Cherry",
    str_detect(sample_name, "_DM[0-9]") ~ "Cherry",
    str_detect(sample_id, "_T[0-9]") ~ "Stonefruit"
  ), type = case_when(
    str_detect(sample_name, "extblank") ~ "Extblank",
    str_detect(sample_name, "SPD$") ~ "SPD",
    str_detect(sample_name, "FF$") ~ "FF",
    TRUE ~ type
  )) %>%
  column_to_rownames("rownames")

taxa_names(ps_run3) <- taxa_names(ps_run3) %>% 
  str_remove("^.*-") %>%
  str_replace("nr.dimidiatus", "truncatus")

ps_run3 <- ps_run3 %>%
  subset_samples(!type == "Extblank")

# Spp summary
spp_stat <- speedyseq::psmelt(ps_run3) %>%
  mutate(comm = case_when(
    type %in% c("ACV", "DC", "FF", "SPD") ~ "Field",
    type %in% c("DrosLarv", "DrosMock") ~ "Mock",
    TRUE ~ type
    )) %>%
  filter(Abundance > 0 ) %>%
  mutate(Genus = case_when(
    str_detect(Genus, "__") ~ as.character(NA),
    TRUE  ~ Species
  )) %>%
  mutate(Species = case_when(
    str_detect(Species, "__") ~ as.character(NA),
    TRUE ~ Species
  )) 

# N ASVs and samples
spp_stat %>%
  group_by(comm) %>%
  summarise(n_extracts = n_distinct(extract_id), n_samples = n_distinct(sample_id), n_asv = n_distinct(OTU), n_spp = n_distinct(Species), n_genus = n_distinct(Genus))


# Spread of Species across different types
spp_stat %>%
  group_by(extract_id, type) %>%
  dplyr::filter(Abundance > 0) %>%
  summarise(counts = n_distinct(Species)) %>%
  ungroup() %>%
  group_by(type) %>%
  summarise(mean = mean(counts), 
            se = sd(counts)/sqrt(length(counts)),
            max = max(counts),
            min = min(counts))

# Number of species for each order
ps_run3 %>%
  speedyseq::psmelt() %>%
  group_by(Order) %>%
  summarise(n = n_distinct(Species))

# Number of species for each genus
ps_run3 %>%
  speedyseq::psmelt() %>%
  group_by(Genus) %>%
  summarise(n = n_distinct(Species)) %>%
  arrange(-n)

# Number of species for each order, seperated by classified
ps_run3 %>%
  speedyseq::psmelt() %>%
  mutate(not_spp = case_when(
    str_detect(Species, "__") ~ TRUE,
    TRUE ~ FALSE
  )) %>%
  group_by(Order, not_spp) %>%
  summarise(n = n_distinct(Species))

```

## Alpha diversity

```{r alpha div}
dir.create("output/alpha")

# Get a histogram of taxon sums
taxa_sums(ps_run3) %>%
  as_tibble(rownames = "OTU") %>%
  filter(value > 0) %>%
  ggplot(aes(x = value)) +
  geom_histogram() 

# Get richness measures
richness <- phyloseq::estimate_richness(ps_run3, measures=c("Chao1", "Shannon", "Simpson")) %>% #
  rownames_to_column("sample_name") %>%
  mutate(sample_name = str_replace_all(sample_name, "\\.", "-"))

# Calculate Faith's PD-index & Species richness - with Standard errors
sespd <- picante::ses.pd(as(phyloseq::otu_table(ps_run3), "matrix"),  phyloseq::phy_tree(ps_run3), null.model = "taxa.labels", include.root = F, runs = 9)

# Join together
div_table <- sespd %>%
  rownames_to_column("sample_name") %>%
  dplyr::select(sample_name, ntaxa, pd.obs) %>%
  dplyr::rename(pd = pd.obs, alpha = ntaxa) %>%
  left_join(richness, by="sample_name") %>%
  left_join(sample_data(ps_run3) %>% 
              as("data.frame") %>%
              filter(!duplicated(sample_id)) %>%
              dplyr::select(amp_rep, sample_name = sample_id, type, orchard),
            by = "sample_name")  

# Filter to just the field collected types
div_table <- div_table %>%
  filter(type %in% c("DC", "ACV", "FF", "SPD"))

# Summarise means + SE
se <- function(x) sqrt(var(x, na.rm=TRUE)/length(x))

div_table %>%
  group_by(type) %>%
  #summarise(across(c("alpha", "pd", "Chao1", "se.chao1", "Shannon", "Simpson"), mean, na.rm=TRUE)) %>%
  summarise(mean = mean(alpha, na.rm=TRUE), se = se(alpha))

# Use breakaway to check that all species have been sufficiently captured
ps_break <- subset_samples(ps_run3, type %in% c("DC", "ACV", "FF", "SPD"))

richness_break <-  ps_break %>% breakaway

break_traps <- summary(richness_break) %>% as_tibble

rich_comparison <- break_traps %>%
  mutate(sample_name = sample_names) %>%
  left_join(div_table) %>%
  mutate(estimate= round(estimate))

table(rich_comparison$estimate == rich_comparison$alpha)
# All true - can go ahead with ANOVAs 

# Difference in alpha diversity between trap types
report::report(aov(alpha ~type, data=div_table))
broom::tidy(TukeyHSD(aov(alpha ~type, data=div_table)))

report::report(aov(Shannon ~type, data=div_table))
broom::tidy(TukeyHSD(aov(Shannon ~type, data=div_table)))

report::report(aov(pd ~type, data=div_table))
broom::tidy(TukeyHSD(aov(pd ~type, data=div_table)))

alpha_results <- bind_rows(
  broom::tidy(TukeyHSD(aov(alpha ~type, data=div_table))) %>% mutate(metric = "richness"),
  broom::tidy(TukeyHSD(aov(Shannon ~type, data=div_table))) %>% mutate(metric = "shannon"),
  broom::tidy(TukeyHSD(aov(pd ~type, data=div_table))) %>% mutate(metric = "phylogenetic")
)

write_csv(alpha_results, "output/alpha/alpha_anova.csv")

# Differences in orchards
report::report(aov(alpha ~orchard, data=div_table))
broom::tidy(TukeyHSD(aov(alpha ~orchard, data=div_table)))

report::report(aov(Shannon ~orchard, data=div_table))
broom::tidy(TukeyHSD(aov(Shannon ~orchard, data=div_table)))

report::report(aov(pd ~orchard, data=div_table))
broom::tidy(TukeyHSD(aov(pd ~orchard, data=div_table)))

# plot differences in alpha diversity

# plot differences in alpha diversity
alpha_plotdat <- div_table %>% 
  pivot_longer(2:7,
               names_to="metric",
               values_to="value") %>%
  filter(metric %in% c("alpha", "Shannon")) %>%
  mutate(metric = metric %>% 
           str_replace("alpha", "Species Richness") %>%
           str_replace("pd", "Phylogenetic Diversity") %>%
           str_replace("Shannon", "Shannon Index")) %>%
  mutate(type = case_when(
    type == "DC" ~ "Syn",
    TRUE ~ type
  )) 

gg.alpha_rich <- alpha_plotdat %>%
  filter(metric == "Species Richness") %>%
  mutate(type = factor(type, levels = c("ACV", "Syn", "SPD", "FF"))) %>%
  ggplot(aes(x=type, y=value,colour=type, fill=type)) +
  geom_boxplot(width =0.9, alpha=0.7, outlier.colour = NA, colour="black")+
  geom_point(position=position_jitter(width = 0.15, height=0),alpha=0.8, shape=21, colour="black")+
  base_theme+
  scale_fill_brewer(palette="Paired") +
  scale_colour_brewer(palette="Paired") +
  facet_grid(.~orchard)  +
  theme(strip.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank()) +
  labs(y = "Species Richness")

gg.alpha_shan <- alpha_plotdat %>%
  filter(metric == "Shannon Index") %>%
  mutate(type = factor(type, levels = c("ACV", "Syn", "SPD", "FF"))) %>%
  ggplot(aes(x=type, y=value,colour=type, fill=type)) +
  geom_boxplot(width =0.9, alpha=0.7, outlier.colour = NA, colour="black")+
  geom_point(position=position_jitter(width = 0.15, height=0),alpha=0.8, shape=21, colour="black")+
  base_theme+
  scale_fill_brewer(palette="Paired") +
  scale_colour_brewer(palette="Paired") +
  facet_grid(.~orchard)  +
  theme(strip.text.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())+
  labs(y = "Shannon Index")

gg.alpha <- gg.alpha_rich / gg.alpha_shan

pdf(file="fig/supplementary/trap_catches.pdf", width = 8, height = 6 , paper="a4r")
  plot(gg.trapweek)
  plot(gg.trapall)
  plot(gg.alpha)
try(dev.off(), silent=TRUE)

```

## Beta diversity 

### Create distance matrix

```{r Distances}
#Subset to remove mocks
ps_traps <- ps_run3 %>%
  subset_samples(!type %in% c("DrosMock", "DrosLarv", "ACV")) %>%
  filter_taxa( function(x) mean(x) > 0, TRUE)

saveRDS(ps_traps, "output/rds/ps_traps.rds")

# Get OTU tables
otutab <- ps_traps %>%
  otu_table() %>%
  as("matrix")

colSums(otutab) > 0

#Impute zeroes for compositional distances
otutab_n0 <- as.matrix(zCompositions::cmultRepl(otutab, method="CZM", output="p-counts"))
#Root phylogenetic tree
phy_tree(ps_traps) <- multi2di(phy_tree(ps_traps))
phy_tree(ps_traps) <- makeNodeLabel(phy_tree(ps_traps), method="number", prefix='n')
name.balance(phy_tree(ps_traps), tax_table(ps_traps), 'n1') #Get root

#Calculate different distance metrics
distlist <- vector("list")

distlist$Jaccard <- as.matrix(vegdist(otutab, method="jac",binary = T))
distlist$Bray <- as.matrix(vegdist(otutab, method="bray"))
distlist$Aitchison <- as.matrix(vegdist(CoDaSeq::codaSeq.clr(otutab_n0), method="euclidean"))
distlist$Philr <- as.matrix(vegdist(philr::philr(otutab_n0, phy_tree(ps_traps),
                                                part.weights='enorm.x.gm.counts',
                                                ilr.weights='blw.sqrt'), method="euclidean"))
distlist$Unifrac <- as.matrix(phyloseq::UniFrac(ps_traps, weighted=FALSE, parallel = TRUE))
distlist$WUnifrac <- as.matrix(phyloseq::UniFrac(ps_traps, weighted=TRUE, parallel = TRUE))

```

### Adonis and betadisper

```{r betatest}
# Adonis test
metadata <- sample_data(ps_traps) %>%
  as("data.frame")%>%
  mutate(orchard_type = paste0(orchard, "_", type)) %>%
  dplyr::select(sample_id, orchard, type, orchard_type)

# Test difference by community type
adonis_results <- distlist %>%
  purrr::map(function(x) {
    y <- as.dist(x[metadata$sample_id, metadata$sample_id])
    bind_rows(
    broom::tidy(adonis(y~type, method="euclidean", data=metadata, permutations = 999)$aov.tab) %>% mutate(comparison = "type"),
    broom::tidy(adonis(y~orchard, method="euclidean", data=metadata, permutations = 999)$aov.tab) %>% mutate(comparison = "orchard"),
    broom::tidy(adonis(y~orchard+type, method="euclidean", data=metadata, permutations = 999)$aov.tab) %>% mutate(comparison = "orchard_type"),
    broom::tidy(adonis(y~orchard*type, method="euclidean", data=metadata, permutations = 999)$aov.tab) %>% mutate(comparison = "orchard_type2")
    )
})  %>%
  bind_rows(.id="dist")

# Check homogeneity
betadisper_results <- distlist %>%
  purrr::map(function(x) {
    y <- as.dist(x[metadata$sample_id, metadata$sample_id])
  bind_rows(
    as.data.frame(permutest(vegan::betadisper(y, metadata$type))$tab) %>%
      mutate(term="type"),
    as.data.frame(permutest(vegan::betadisper(y, metadata$orchard))$tab) %>%
      mutate(term="orchard"),
    as.data.frame(permutest(vegan::betadisper(y, metadata$orchard_type))$tab) %>%
      mutate(term="orchard_type")
  )
})  %>%
  bind_rows(.id="dist")

dir.create("output/beta")
write_csv(adonis_results, "output/beta/adonis.csv")
write_csv(betadisper_results, "output/beta/adonis.csv")

#Plot betadispersal
z <- as.dist(distlist$WUnifrac[metadata$sample_id, metadata$sample_id])
orchard_dispnest <- vegan::betadisper(z, metadata$orchard)
plot(orchard_dispnest, ellipse = TRUE, hull = FALSE)

sample_dispnest <- vegan::betadisper(z, metadata$type)
plot(sample_dispnest, ellipse = TRUE, hull = FALSE)

orchard_type_dispnest <- vegan::betadisper(z, metadata$orchard_type)
plot(orchard_type_dispnest, ellipse = TRUE, hull = FALSE)

```

### MVAbund

```{r mvabund}
# Get species occurance table
otutab <- otu_table(ps_traps) %>%
  as.matrix() %>%
  as.data.frame()

# Get metadata 
metadata <- sample_data(ps_traps) %>%
  as("data.frame")%>%
  mutate(orchard_type = paste0(orchard, "_", type)) %>%
  dplyr::select(sample_id, orchard, type, orchard_type)


spp_occur <- mvabund(otutab)

# Mean - variance relationship
meanvar.plot(spp_occur)

model_res <- vector("list", length=4)

# Model 1 - trap type
model_res[[1]] <- manyglm(spp_occur ~ metadata$type, family="negative.binomial")

# Model 2 - orchard
model_res[[2]] <- manyglm(spp_occur ~ metadata$orchard, family="negative.binomial")

# Model 3 - orchard + type
model_res[[3]] <- manyglm(spp_occur ~ metadata$orchard+metadata$type, family="negative.binomial")

# Model 4 - orchard * type
model_res[[4]] <- manyglm(spp_occur ~ metadata$orchard*metadata$type, family="negative.binomial")

saveRDS(model_res, "model_res.rds")

# Run multivariate anovas
anova_res <- vector("list", length=length(model_res))

anova_res[[1]] <- anova(model_res[[1]])
anova_res[[2]] <- anova(model_res[[2]])
anova_res[[3]] <- anova(model_res[[3]])
anova_res[[4]] <- anova(model_res[[4]])

saveRDS(anova_res, "anova_res.rds")

# run univariate tests 
univa_res <- vector("list", length=length(model_res))

univa_res[[1]] <- anova(model_res[[1]], p.uni="adjusted")
univa_res[[2]] <- anova(model_res[[2]], p.uni="adjusted")
univa_res[[3]] <- anova(model_res[[3]], p.uni="adjusted")
univa_res[[4]] <- anova(model_res[[4]], p.uni="adjusted")

saveRDS(univa_res, "univa_res.rds")

# Read in manyglm results
model_res <- readRDS("output/mvabund/model_res.rds")
anova_res <- readRDS("output/mvabund/anova_res.rds")
univa_res <- readRDS("output/mvabund/univa_res.rds")

# Check which model has the highest lowest AIC
model_res %>%
  purrr::map(~pluck(.x,"AICsum")) %>%
  unlist()

# Use model 3 as we gain use of both factors without substantially increasing AIC
anova_res[[3]]$table

# Check for significant interaction effect in model 4
anova_res[[4]]$table

```

### PCoA plots

```{r pca plots}
# Get distance
plot_dist <- as.dist(distlist$WUnifrac)

plot_pcoa <- ordinate(ps_traps, 'PCoA', distance=plot_dist)

sample_data(ps_traps) <- sample_data(ps_traps) %>%
  as("matrix") %>%
  as_tibble(rownames="sample") %>%
  mutate(type = type %>% str_replace("DC", "Syn")) %>%
  mutate(orchard_type = paste0(orchard, "_", type)) %>%
  column_to_rownames("sample")
  
gg.pca <- plot_ordination(ps_traps, plot_pcoa, color = "type", shape="orchard") +
  #stat_ellipse(geom = "polygon", aes(group = orchard, fill=orchard, colour=orchard), alpha=0.05) +
  scale_fill_viridis_d()+
  scale_colour_viridis_d()+
  new_scale_fill() +
  new_scale_colour() +
  geom_point(size=4, alpha=1, aes(colour=type, shape=orchard)) +
  #stat_ellipse(geom = "polygon", aes(group = type, fill=type, colour=type), alpha=0.05) +
  scale_colour_brewer(palette="Paired") +
  scale_fill_brewer(palette="Paired") +
  geom_hline(yintercept = 0, linetype=2, alpha=0.5) +  
  geom_vline(xintercept = 0, linetype=2, alpha=0.5) +
  base_theme +
  #coord_fixed() +
  theme(legend.position = "bottom") +
  labs(colour = "Sample Type") 

gg.pca

# Heirarchial clustering
dend <- as.phylo(hclust(plot_dist, method="ward.D2"))

dend$tip.label <- dend$tip.label %>%
  str_remove("HLVKYDMXX_")
p3 <- ggtree(dend) + 
  theme_tree2()

colours_p3 <- p3$data %>%
  left_join(as_data_frame(sample_data(ps_traps)) %>%
              dplyr::select(sample_id, type, orchard) %>%
  dplyr::rename(label = sample_id) %>%
    mutate(label = label %>%
  str_remove("HLVKYDMXX_"))
    )

p3 <- p3 %<+% colours_p3  + 
  geom_tippoint(aes(colour=as.factor(type), shape=as.factor(orchard)), size=3)  + 
  geom_tiplab(aes(colour=as.factor(type)))+
  scale_colour_brewer(palette="Paired") +
  theme_void()+
  scale_x_continuous(expand=c(0, 0.3)) +
  theme(legend.position = "none")

# Plot together
gg.pca_tree <- p3 + gg.pca + plot_layout(widths = c(1,2)) + plot_annotation(tag_levels = "A")

gg.pca_tree

# Write out PCA plots
pdf(file="fig/supplementary/pca_tree.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg.pca_tree)
try(dev.off(), silent=TRUE)
```


# Phylogeny of trap catches

```{r Phylogeny}
tree <- phy_tree(ps_run3)

# plot 
plot_dat <- speedyseq::psmelt(ps_run3) %>%
  mutate(orchard = case_when(
    str_detect(sample_name,"^T") ~ "Stonefruit",
    str_detect(sample_name,"^M") ~ "Cherry",
    TRUE ~ as.character(NA)
  )) %>%
    dplyr::filter(type %in% c("ACV", "DC", "FF", "SPD")) %>%
  mutate(type = case_when(
    
    type == "DC" ~ "Syn",
    TRUE ~ type
  )) %>%
  mutate(type = factor(type, levels = c("ACV", "Syn", "SPD", "FF"))) %>%
  filter(!str_detect(OTU, "__")) 

p1dat <- plot_dat %>%
  mutate(Abundance = case_when(
    orchard == "Stonefruit" ~ 0,
    TRUE ~ Abundance
    )) %>%
  dplyr::select(sample_id, OTU, type, Abundance) %>%
  group_by(sample_id) %>%  
  mutate_at(vars(Abundance), ~ . / sum(.)) %>%
  group_by(OTU, type) %>%
  summarise(Abundance = mean(Abundance, na.rm = TRUE)) %>%
  distinct() %>%
  pivot_wider(names_from = type,
              values_from = Abundance,
              values_fill = 0) %>%
  column_to_rownames("OTU") 

tree <- drop.tip( , tree$tip.label[!tree$tip.label %in% plot_dat$OTU])

circ <- ggtree(tree, layout = "circular", aes(color=Order))+
  geom_tippoint()+
  geom_tiplab(offset=0.7, align=TRUE, size=2.5, colour="black")

circ_dat <- circ$data %>%
  left_join(speedyseq::psmelt(ps_run3)  %>%
            dplyr::select(label = OTU, Order) %>%
              distinct() 
    )

circ <- circ %<+% circ_dat +
  scale_colour_brewer(palette="Paired",na.value="black") 


p1 <- gheatmap(circ, p1dat, offset=0, width=.3, color="gray20",
               colnames_angle=95, colnames_offset_y = 0) +
    scale_fill_viridis_c(option="B", name="Cherry Abundance", alpha=0.9, trans="log10", label= scales::percent) 

p2dat <- plot_dat %>%
  mutate(Abundance = case_when(
    orchard == "Cherry" ~ 0,
    TRUE ~ Abundance
    )) %>%
  dplyr::select(sample_id, OTU, type, Abundance) %>%
  group_by(sample_id) %>%  
  mutate_at(vars(Abundance), ~ . / sum(.)) %>%
  group_by(OTU, type) %>%
  summarise(Abundance = mean(Abundance, na.rm = TRUE)) %>%
  distinct() %>%
  pivot_wider(names_from = type,
              values_from = Abundance,
              values_fill = 0) %>%
  column_to_rownames("OTU") 

  
p2 <- p1 + new_scale_fill()
gg.phylo <- gheatmap(p2, p2dat, offset=0.34, width=.3,
         colnames_angle=90, colnames_offset_y = .1, color="gray20") +
    scale_fill_viridis_c(option="D", name="Stonefruit Abundance", trans="log10", label= scales::percent) +
theme(legend.position = "bottom")

gg.phylo


# FInal phylo fig
gg.Fig4a <- (gg.trapall + theme(strip.text.y = element_blank(), axis.title.y= element_blank(),
                                )) / gg.alpha_rich / gg.alpha_shan 
gg.Fig4 <- gg.phylo - gg.Fig4a + plot_layout(widths = c(2,1.5)) + plot_annotation(tag_levels = "A")

gg.Fig4

# Write out bias explained for supplementary 
pdf(file="fig/Fig4_diversity.pdf", width = 11, height = 8 , paper="a4r")
  plot(gg.Fig4)
try(dev.off(), silent=TRUE)
  
```

# Check presence of taxa in Australia
```{r}
# Check presence on AFD
afd_check <- ps2 %>%
  speedyseq::psmelt() %>%
  filter(!str_detect(Species, "Synthetic")) %>%
  dplyr::filter(type %in% c("ACV", "DC", "FF", "SPD")) %>%
  dplyr::group_by(Family, Genus, Species) %>%
  summarise(metabarcoding_reads = sum(Abundance)) %>%
  filter(metabarcoding_reads > 0) %>%
  filter(!str_detect(Species, "__")) %>%
  mutate(Species = Species %>% str_replace_all("_", " ")) %>%
  mutate(
    Family_present = afdscraper::check_afd_presence(Family),
    Genus_present = afdscraper::check_afd_presence(Genus),
    Species_present = afdscraper::check_afd_presence(Species)
  ) %>%
  dplyr::select(Family, Family_present, Genus, Genus_present, 
                Species, Species_present, metabarcoding_reads)
    
write_csv(afd_check, "output/afd_check.csv")

# Check presence on ALA
# First we need to set some data quality filters for ALA
# To view available filters, run: find_field_values("basis_of_record")
ala_quality_filter <- galah::select_filters(
      basisOfRecord = c("PreservedSpecimen", "LivingSpecimen",
                      "MaterialSample", "NomenclaturalChecklist"),
      profile = "ALA")

ala_check <- ps2 %>%
  speedyseq::psmelt() %>%
  filter(!str_detect(Species, "Synthetic")) %>%
  dplyr::filter(type %in% c("ACV", "DC", "FF", "SPD")) %>%
  dplyr::group_by(Family, Genus, Species) %>%
  summarise(metabarcoding_reads = sum(Abundance)) %>%
    filter(metabarcoding_reads > 0) %>%
  filter(!str_detect(Species, "__")) %>%
  mutate(Species = Species %>% str_replace_all("_", " ")) %>%
  mutate(
    species_present = purrr::map(Species, function(x){
    # first check name
    query <- select_taxa(x) %>% 
      as_tibble()%>%
      dplyr::filter(across(any_of("match_type"), ~!.x == "higherMatch"))
    # Then get occurance counts
    if(!is.null(query$scientific_name)){
      ala_occur <- ala_counts(taxa=query, filters=ala_quality_filter)
      return(data.frame(Species_present = ifelse(ala_occur > 0, TRUE, FALSE), ALA_counts = ala_occur))
    } else {
      return(data.frame(Species_present = FALSE, ALA_counts = 0))
    }
    })) %>%
  unnest(species_present) %>%
  dplyr::select(Family, Genus, Species, Species_present, ALA_counts, metabarcoding_reads)

write_csv(ala_check, "output/ala_check.csv")

```